### PR TITLE
Integrate TaskManager into schema migration demo

### PR DESCRIPTION
**Pull request by scottn12**
## Description

This PR integrates `TaskManager` into the migration flow of the schema upgrade demo. Previously, all clients would race to migrate the container while checking to see if another client has completed it first. This PR changes that by only allowing one client to execute the migration with `TaskManager`. 

Additionally, this PR helps support retry scenarios by storing some of the migration work.

## Reviewer Guidance

Would appreciate feedback on:
- Implementation details for this specific demo.
- General thoughts on `TaskManager` APIs and coding patterns.
- If we need to close `this._preparedDetachedModel` in scenarios where the migration is completed by another client.
- How we enter the state machine. Specifically, how we re-enter the state machine in the case of disconnection during the migration flow.

## Other information or known dependencies

[AB#324](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/324)
[AB#1918](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1918)

### CODE REVIEW

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 1]
```diff
@@ -172,11 +153,39 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+            migratedModel?.close();
+            this._migrationP = undefined;
             this.takeAppropriateActionForCurrentMigratable();
         };
 
+        const onAssigned = (taskId: string) => {
+            if (taskId !== this.migrationTaskId) {
+                return;
+            }
+            this._migrationP = doTheMigration().catch(console.error);
```
Comment (scottn12) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:180]:
```suggestion
            this._migrationP = doTheMigration().catch((error) => {
                // Allow next client to retry if we fail
                console.error(error);
                this.taskManager.abandon(this.migrationTaskId);
            });
```

@ChumpChief, do you think this would be sufficient to handle retry scenarios? Basically, if we error out of the migration then we would give up the task and let the next client in queue try. We would also automatically re-enter the queue because we subscribed to the task. Let me know what you think!

File: examples/hosts/app-integration/schema-upgrade/src/migrationInterfaces/migrationTool.ts [hunk 2]
```diff
@@ -34,6 +35,11 @@ export interface IMigrationTool extends IEventProvider<IMigrationToolEvents> {
      */
     finalizeMigration(id: string): Promise<void>;
 
+    /**
+     * TaskManager DDS to facilitate the migration execution.
+     */
+    readonly taskManager: ITaskManager;
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrationInterfaces/migrationTool.ts:41]: This interface generally doesn't reveal the DDSs that can be used for migration, but instead provides special-purpose API surface for migration while keeping them concealed.  Maybe we want to do that here too (I have not thought through in any detail though)?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 3]
```diff
@@ -131,39 +139,12 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
-
             // TODO: Support retry
-            // TODO: Use TaskManager here to reduce container noise.  Specifically -- all clients should race up to
-            // this point (so they're all prepared to do the migration) but should wait for the task lock before
-            // attempting the attach() to minimize the chance that multiple containers are created on the service.
             const containerId = await createResponse.attach();
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:143]: It's OK for all clients to simultaneously do everything up until the attach call.  However, since it's currently all in doTheMigration, clients will only do the export/transform/import-into-detached once they get the task, which will be slower if a second client needs to take over.  Consider splitting this into two chunks - preparing for the migration (not guarded by TaskManager) vs. attach/finalize (guarded by task manager)

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 4]
```diff
@@ -172,11 +153,39 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+        };
+
+        const exitMigration = (taskId: string) => {
+            if (taskId !== this.migrationTaskId) {
+                return;
+            }
+            this.taskManager.off("lost", exitMigration);
+            this.taskManager.off("completed", exitMigration);
+            this.taskManager.off("assigned", onAssigned);
+            migratedModel?.close();
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:171]: I don't think it's necessary to close the migratedModel here (at least in all cases) - e.g. if I lose connection for a second and so lose the task, the migratedModel I'm holding is still good.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 5]
```diff
@@ -131,39 +139,12 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-            // above await.
-            if (migratable.migrationTool.migrationState as MigrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
-
             // TODO: Support retry
             await migratable.migrationTool.finalizeMigration(containerId);
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:146]: Suppose I successfully attach(), but then lose the task before I can finalizeMigration().  If I get the task again, I'll attach() another new container currently -- it would be better if we can retain that already-attached container and just call finalizeMigration().

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 1 above)
Comment (ChumpChief) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:180]: I put a couple notes above about reusing attached containers and such - although looking to other clients to pick up the task may work OK, probably consider a retry path for the local client to minimize extra containers getting attached and to be a little faster.  This is most true in the scenario that a client is the only one connected to the collaboration session - abandoning and re-volunteering introduces a delay that we can probably avoid.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 5 above)
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:146]: Also, if I exitMigration during attach(), nothing will notify this codepath that it should stop and not call finalizeMigration().

**Review by ChumpChief — COMMENTED** [review]
I think driving the flow off the events can make a lot of sense for structure, but probably want to re-slice the doTheMigration to appropriate granularity then.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 6]
```diff
@@ -86,97 +95,150 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
         }
 
         const doTheMigration = async () => {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:97]: Maybe add a comment either here or on ensureMigrating (or both) explaining the contract/calling patterns.  E.g. "we assume ensureMigrating will only be called once and resolve only after the migration is complete" vs. "ensureMigrating is called to start the migration, but also can be called if the migration is interrupted and needs to be reattempted (in which case it will have this state...)".  Bonus if these calling requirements can be validated/enforced with asserts.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 7]
```diff
@@ -86,97 +95,150 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+
+                // Store the detached model for later use and retry scenarios
+                this._detachedModel = detachedModel;
+            };
+
+            const completeTheMigration = async () => {
+                assert(this._detachedModel !== undefined, "this._detachedModel should be defined");
+
+                if (this._containerId === undefined) {
+                    this._containerId = await this._detachedModel.attach();
```
[the remark this answers is not in the document]
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts]: To clarify, are you suggesting moving the `volunteerForTask()` inside `completeTheMigration()`?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 8]
```diff
@@ -86,97 +95,150 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
             }
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+            // Volunteer to complete the migration.
+            let isAssigned: boolean;
+            try {
+                isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
+            } catch (error) {
+                // If we error here either the container was closed by another client (during the migration process),
```
[the remark this answers is not in the document]
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts]: On more thought I think what causing the error is a race between receiving the `complete` op from TaskManager and this block in `migrationTool.ts`:
```ts
this.crc.on("atomicChanged", (key: string) => {
    if (key === newContainerIdKey) {
        this.emit("migrated");
    }
});
```
If the above block fires first, then we will emit migrated and close the container on the local client before `volunteerForTask()` resolves (due to the assigned client calling `complete()`). I'll try another approach in the next commit!

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 9]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-            } else if (this.dataTransformationCallback !== undefined) {
-                // Otherwise, try using the dataTransformationCallback if provided to get the exported data into
-                // a format that we can import.
-                try {
-                    transformedData = await this.dataTransformationCallback(exportedData, migratedModel.version);
-                } catch {
-                    // TODO: This implies that the contract is to throw if the data can't be transformed, which isn't
-                    // great.  How should the dataTransformationCallback indicate failure?
+            const prepareTheMigration = async () => {
+                if (this._preparedDetachedModel !== undefined) {
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:116]: Who ensures that every client will land on the same detached model, without any data loss? Is it modelLoader? I'm curious what are some of the contraints there?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 10]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
         }
 
         const doTheMigration = async () => {
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:110]: General q (maybe not directly tied to this PR) - is user expected to write all this logic to handle their own use cases? Would it be helpful to export this as experimental "migration" package where user can consume simplified set of APIs, without needing to understand all the internal contraints and various concepts/DDS-es we rely on to achieve migration?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 11]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-            // TODO: Does the app developer have everything they need to dispose gracefully when recovering with a new
-            // ModelLoader?
-            const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
-            if (!migrationSupported) {
-                this.emit("migrationNotSupported", acceptedVersion);
-                this._migrationP = undefined;
-                return;
-            }
+            // doTheMigration() is called at the start of migration and should only resolve in two cases. First, is if
+            // we complete the migration. Second, is if we disconnect during the migration process. In both cases we
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:112]: "Complete" means do actual migration or exit without completing because another client completed it ?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 12]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                // It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.
+                // Currently this fails the migration gracefully and emits an event so the app developer can know
+                // they're stuck. Ideally the app developer would find a way to acquire a new ModelLoader and move
+                // forward, or at least advise the end user to refresh the page or something.
+                // TODO: Does the app developer have everything they need to dispose gracefully when recovering with
+                // a new ModelLoader?
+                const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
+                if (!migrationSupported) {
                     this.emit("migrationNotSupported", acceptedVersion);
                     this._migrationP = undefined;
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:130]: Assignment of _migrationP seems a bit odd and hard to follow. Wouldn't it get cleared when doTheMIgration completes anyways?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 13]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                    return;
+                }
+
+                // It's possible that our modelLoader is older and doesn't understand the new acceptedVersion.
+                // Currently this fails the migration gracefully and emits an event so the app developer can know
+                // they're stuck. Ideally the app developer would find a way to acquire a new ModelLoader and move
+                // forward, or at least advise the end user to refresh the page or something.
+                // TODO: Does the app developer have everything they need to dispose gracefully when recovering with
+                // a new ModelLoader?
+                const migrationSupported = await this.modelLoader.supportsVersion(acceptedVersion);
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:127]: Do we have mechanism where all connected clients can align on proposed "upgrade" version? Otherwise, if one connected client cannot upgrade, other connected clients may not wait before upgrading schema?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 14]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                    return;
+                }
 
-            this.takeAppropriateActionForCurrentMigratable();
+                await migratable.migrationTool.finalizeMigration(this._containerId);
+
+                this.currentModel.migrationTool.completeMigrationTask();
+            };
+
+            await prepareTheMigration();
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:207]: You could be more explicit what f does: this._detachedModel = await prepareDetachedModel(); and bail here if model is not available.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 15]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                } else if (this.dataTransformationCallback !== undefined) {
+                    // Otherwise, try using the dataTransformationCallback if provided to get the exported data into
+                    // a format that we can import.
+                    try {
+                        transformedData = await this.dataTransformationCallback(exportedData, migratedModel.version);
+                    } catch {
+                        // TODO: This implies that the contract is to throw if the data can't be transformed, which
+                        // isn't great.  How should the dataTransformationCallback indicate failure?
+                        this.emit("migrationNotSupported", acceptedVersion);
+                        this._migrationP = undefined;
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:156]: See other comment above around this._migrationP

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 16]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-            }
+            const completeTheMigration = async () => {
+                assert(this._preparedDetachedModel !== undefined, "this._preparedDetachedModel should be defined");
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                // Volunteer to complete the migration.
+                let isAssigned: boolean;
+                try {
+                    isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:177]: Does volunteerForMigration acquires lock?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 17]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
-            // container/model to use.  There could maybe be some efficiency gain by retaining the model in the
-            // case that we win the race?  But it likely just doesn't matter that much because the Loader probably
-            // cached the Container anyway.
+                if (this._containerId === undefined) {
+                    this._containerId = await this._preparedDetachedModel.attach();
+                }
 
-            this._migrationP = undefined;
+                // Check to make sure we still have the task assignment.
+                if (!this.currentModel.migrationTool.haveMigrationTask()) {
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:197]: Would you have to do this after every async call, while you have lock? As we chatted it may be helpful to have some taskmanager helper function that helps you manage the flow.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 18]
```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                    // Exit early if we lost the task assignment, we are most likely disconnected.
+                    return;
+                }
 
-            this.takeAppropriateActionForCurrentMigratable();
+                await migratable.migrationTool.finalizeMigration(this._containerId);
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:202]: What would play out here if you lose the lock?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 19]
```diff
@@ -60,6 +73,12 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
      */
     private readonly takeAppropriateActionForCurrentMigratable = () => {
+        if (!this.connected) {
+            // If we are not connected we should wait until we reconnect and try again.
+            this.currentModel.migrationTool.once("connected", this.takeAppropriateActionForCurrentMigratable);
+            return;
+        }
+
         const migrationState = this._currentModel.migrationTool.migrationState;
         if (migrationState === "migrating") {
```
Comment (ssimic2) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:83]: If this is the state machine, some comments would be helpful what each state should be representing. "migrating" can mean "im migrating" or "someone is migrating". Some comments around the "else" case as well.

**Comment by ssimic2** [comment]
Some general thoughts:

1. It would be nice if we had "migration" package that essentially hides all this complexity of variations and unhappy paths that can play out with clear success/fail outcomes.
2. It would be nice to understand use cases out migration should support. Then it may be easier to reason around success/fail outcomes. For example if we have use case scenario where user wants migration to occur only if certian condition was met (say all clients can upgrade), we could reason what building blocks we need in place.
3. From perspective of TaskManager it would be nice to abstract some helper functions that know how to handle various error cases as they execute async block.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 19 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:83]: There are comments in `migrationInterfaces/migrationTool.ts` which explain the migration states. Do you think we should bring it to migrator.ts as well to increase visibility?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 10 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:110]: I'm not sure what the long term goals are here. Maybe @ChumpChief has a better idea?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 11 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:112]: It can mean either in this case. It just means that we exited after either the local client or another client has successfully completed the client. I can update the comment to clarify!

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 9 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:116]: That is what `completeMigration()` takes care of. That function will ensure a single client (through TaskManager) attaches their local detached model and then shares it with the other clients

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 13 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:127]: Yes, that logic is handled by the `MigrationTool` class with the proposed and accepted version mechanism that Matt implemented. I haven't really touched that logic in this PR though

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 12 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:130]: Yeah it should get cleared. Upon a second look, I also noticed we would hit the assert error below (since we are exiting for a reason other than completing the migration succesfully or disconnecting), so we probably need to update this mechanism as well.
```ts
assert(this.currentModel.migrationTool.newContainerId !== undefined,
                        "newContainerId should be defined");
```

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 16 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:177]: It will try to acquire the lock yes. If it does then `isAssigned` will be true, and if not then it will return false (another client completed it), or we will throw an error due to a disconnection.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 17 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:197]: That is correct. I decided to keep the check here since this is a relatively small task. If the task was larger with more async calls, I would probably opt to take a different approach that relies on listening to the `"lost"` event on TaskManager to determine if we need to exit the flow.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 18 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:202]: We would end up throwing an error on the next line (`completeMigraitonTask()`), since we are not able to complete a task without being assigned. I think we should probably try to prevent that with another check though

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 11 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:112]: Feel free to resolve this comments as you make clarifications.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 12 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:130]: Feel free to resolve these comments as you make the changes.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 10 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:110]: I'll resolve the convo since it's not related to this PR, and we move it to general category questions.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 17 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:197]: OK, I see. That makes sense. I would say, these examples often become reference usage for various concepts. Would promise race between "lost" event and async task be considered good simple general pattern? If so, would that be appropriate solution to incorporate here? Given this is simple use case, I'll leave it up to you. Feel free to resolve the conversation with whichever direction you conclude on.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 18 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:202]: As suggested above, would promise race be a general pattern that helps us avoid these checks?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 19 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:83]: " * * migrating - a proposal to migrate has been accepted.  The data is currently being migrated.
I could not understand what it meant in this particular context.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 9 above)
Comment (ssimic2) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:116]: As I understood completeMigration completes the migration of the detached mode? My question was mostly who ensures that "detached" model

> Some general thoughts:
> 
> 1. It would be nice if we had "migration" package that essentially hides all this complexity of variations and unhappy paths that can play out with clear success/fail outcomes.
> 2. It would be nice to understand use cases out migration should support. Then it may be easier to reason around success/fail outcomes. For example if we have use case scenario where user wants migration to occur only if certian condition was met (say all clients can upgrade), we could reason what building blocks we need in place.
> 3. From perspective of TaskManager it would be nice to abstract some helper functions that know how to handle various error cases as they execute async block.

Btw these are general comments going forward, I understand that scope of this PR is mostly focused on TaskManager for migration.

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 20]
```diff
@@ -60,6 +73,12 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
      * that a freshly-loaded migrated container is in collaborating state.
      */
     private readonly takeAppropriateActionForCurrentMigratable = () => {
+        if (!this.connected) {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:76]: We don't really have to be connected to `ensureLoading()` -- maybe this should be scoped to `ensureMigrating()`?

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts [hunk 21]
```diff
@@ -86,100 +108,132 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
...
+                    // re-enter the state machine which will wait until we reconnect.
+                    // Note: while we wait to reconnect it is possible that another client will have already completed
+                    // the migration.
+                    assert(!this.connected, "We should be disconnected");
+                    return;
+                }
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                if (!isAssigned) {
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:183]: It's possible that another client successfully finalized the migration (wrote the new ID to the CRC) but disconnected/died before they were able to send the "complete" op -- if this happens we should avoid proceeding here.  I think this means we should check the migrationTool.newContainerId rather than trusting the isAssigned result here.

**Review by ChumpChief — COMMENTED** [review]
The flow is looking good!  I see a couple small gotchas but I think these are probably the last ones :)

File: examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts (hunk 21 above)
Comment (scottn12) [reply examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts:183]: Makes sense. We should also validate, probably with an assert, that `isAssigned` is true if we make it past that check

**Review by ChumpChief — APPROVED** [review]
Nice work :) Excited to have this in!
